plugins/modules/dcnm_image_upgrade.py

```diff
@@ -0,0 +1,5313 @@
+        self.validated = copy.deepcopy(valid_params)
+
+        if invalid_params:
+            msg = "Invalid parameters in playbook: "
+            msg += f"{','.join(invalid_params)}"
+            self.module.fail_json(msg)
+
+    def _validate_input_for_query_state(self):
+        """
+        Caller: self.validate_input()
+
+        Validate that self.config contains appropriate values for query state
+
+        NOTES:
+        1. This is currently identical to _validate_input_for_merged_state()
+        2. Adding in case there are differences in the future
+        """
+        params_spec = self._build_params_spec_for_merged_state()
+        if not self.config:
+            msg = "config: element is mandatory for state query"
+            self.module.fail_json(msg)
+
+        valid_params, invalid_params = validate_list_of_dicts(
+            self.config.get("switches"), params_spec, self.module
+        )
+        # We're not using self.validated. Keeping this to avoid
+        # linter error due to non-use of valid_params
+        self.validated = copy.deepcopy(valid_params)
+
+        if invalid_params:
+            msg = "Invalid parameters in playbook: "
+            msg += f"{','.join(invalid_params)}"
+            self.module.fail_json(msg)
+
+    def _merge_global_and_switch_configs(self, config):
+        """
+        Merge the global config with each switch config and return
+        a dict of switch configs keyed on switch ip_address.
+
+        Merge rules:
+        1.  switch_config takes precedence over global_config.
+        2.  If switch_config is missing a parameter, use parameter
+            from global_config.
+        3.  If a switch_config has a parameter, use it.
+        4.  If global_config and switch_config are both missing an
+            optional parameter, use the parameter's default value.
+        5.  If global_config and switch_config are both missing a
+            mandatory parameter, fail.
+        """
+        if not config.get("switches"):
+            msg = f"{self.class_name}._merge_global_and_switch_configs: "
+            msg += "playbook is missing list of switches"
+            self.module.fail_json(msg)
+
+        msg = f"REMOVE: {self.class_name}._merge_global_and_switch_configs: "
+        msg += f"config: {config}"
+        self.log_msg(msg)
+        global_config = {}
+        global_config["policy"] = config.get("policy")
+        global_config["stage"] = config.get("stage")
+        global_config["upgrade"] = config.get("upgrade")
+        global_config["options"] = config.get("options")
+        global_config["validate"] = config.get("validate")
+        msg = f"REMOVE: {self.class_name}._merge_global_and_switch_configs: "
+        msg += f"global_config: {global_config}"
+        self.log_msg(msg)
+        self.switch_configs = []
+        for switch in config["switches"]:
+            switch_config = global_config.copy() | switch.copy()
```

Comment:
I rewrote the merge logic since "|" was not good enough anyway :-)  Please see the latest commit, merge_dicts(), in module_utils/image_mgmt/image_upgrade_common.py.